proto/src/lib.rs

```diff
@@ -73,10 +73,174 @@ mod tendermint {
 
 pub use tendermint::*;
 
-mod domaintype;
-pub use domaintype::DomainType;
-
 mod error;
+use anomaly::BoxError;
+use bytes::{Buf, BufMut};
 pub use error::{Error, Kind};
+use prost::encoding::encoded_len_varint;
+use prost::Message;
+use std::convert::{TryFrom, TryInto};
 
 pub mod serializers;
+
+/// Allows for easy Google Protocol Buffers encoding and decoding of domain
+/// types with validation.
+///
+/// ## Examples
+///
+/// ```rust
+/// use bytes::BufMut;
+/// use prost::Message;
+/// use std::convert::TryFrom;
+/// use tendermint_proto::Protobuf;
+///
+/// // This struct would ordinarily be automatically generated by prost.
+/// #[derive(Clone, PartialEq, Message)]
+/// pub struct MyRawType {
+///     #[prost(uint64, tag="1")]
+///     pub a: u64,
+///     #[prost(string, tag="2")]
+///     pub b: String,
+/// }
+///
+/// #[derive(Clone)]
+/// pub struct MyDomainType {
+///     a: u64,
+///     b: String,
+/// }
+///
+/// impl MyDomainType {
+///     /// Trivial constructor with basic validation logic.
+///     pub fn new(a: u64, b: String) -> Result<Self, String> {
+///         if a < 1 {
+///             return Err("a must be greater than 0".to_owned());
+///         }
+///         Ok(Self { a, b })
+///     }
+/// }
+///
+/// impl TryFrom<MyRawType> for MyDomainType {
+///     type Error = String;
+///
+///     fn try_from(value: MyRawType) -> Result<Self, Self::Error> {
+///         Self::new(value.a, value.b)
+///     }
+/// }
+///
+/// impl From<MyDomainType> for MyRawType {
+///     fn from(value: MyDomainType) -> Self {
+///         Self { a: value.a, b: value.b }
+///     }
+/// }
+///
+/// impl Protobuf<MyRawType> for MyDomainType {}
+///
+///
+/// // Simulate an incoming valid raw message
+/// let valid_raw = MyRawType { a: 1, b: "Hello!".to_owned() };
+/// let mut valid_raw_bytes: Vec<u8> = Vec::new();
+/// valid_raw.encode(&mut valid_raw_bytes).unwrap();
+/// assert!(!valid_raw_bytes.is_empty());
+///
+/// // Try to decode the simulated incoming message
+/// let valid_domain = MyDomainType::decode(valid_raw_bytes.clone().as_ref()).unwrap();
+/// assert_eq!(1, valid_domain.a);
+/// assert_eq!("Hello!".to_owned(), valid_domain.b);
+///
+/// // Encode it to compare the serialized form to what we received
+/// let mut valid_domain_bytes: Vec<u8> = Vec::new();
+/// valid_domain.encode(&mut valid_domain_bytes).unwrap();
+/// assert_eq!(valid_raw_bytes, valid_domain_bytes);
+///
+/// // Simulate an incoming invalid raw message
+/// let invalid_raw = MyRawType { a: 0, b: "Hello!".to_owned() };
+/// let mut invalid_raw_bytes: Vec<u8> = Vec::new();
+/// invalid_raw.encode(&mut invalid_raw_bytes).unwrap();
+///
+/// // We expect a validation error here
+/// assert!(MyDomainType::decode(invalid_raw_bytes.as_ref()).is_err());
+/// ```
+pub trait Protobuf<T: Message + From<Self> + Default>
+where
+    Self: Sized + Clone + TryFrom<T>,
+    <Self as TryFrom<T>>::Error: Into<BoxError>,
+{
+    /// Encodes the DomainType into a buffer.
```

Comment:
Nit:
Maybe we could stop referring to Self as "DomainType" in the comments for these methods. For example:

```suggestion
    /// Encodes the domain type (self) into a buffer.
```

Reply:
Yes you're right, I forgot to check the comments. Will update 👍